### PR TITLE
Fix issue #182

### PR DESCRIPTION
**Pull request by ulanzetz**

### CODE REVIEW

**Review by krzemin — COMMENTED** [review]
Good job, thank you 👍

Would that enable transformations between flat and deep sealed families? If yes, I would add it to the test suite.

Maybe we can update the [docs page](https://scalalandio.github.io/chimney/transformers/customizing-transformers.html#transforming-coproducts) and note explicitly that deep sealed hierarchies are supported?

**Review by ldrygala — APPROVED** [review]
Great job 🚀

**Comment by ulanzetz** [comment]
> Good job, thank you 
> 
> Would that enable transformations between flat and deep sealed families? If yes, I would add it to the test suite.
> 
> Maybe we can update the [docs page](https://scalalandio.github.io/chimney/transformers/customizing-transformers.html#transforming-coproducts) and note explicitly that deep sealed hierarchies are supported?

Added tests and updated docs, but I think it will be unobvious if something like this will be available
```
object From {
  sealed trait Base

  sealed trait BranchA extends Base
  sealed trait BranchB extends Base

  case object FirstLeaf extends BranchA
  case object SecondLeaf extends BranchB
}

object To {
  sealed trait Base

  sealed trait BranchA extends Base
  sealed trait BranchB extends Base

  case object FirstLeaf extends BranchB
  case object SecondLeaf extends BranchA
}

(From.FirstLeaf: From.Base).transformInfo[To.Base] ==> To.FirstLeaf
```
Maybe allow only transformations, when list of destination parents is prefix of list of source parents and order of parents is linear?

File: docs/source/transformers/customizing-transformers.rst [hunk 1]
```diff
@@ -173,3 +173,6 @@ do that. Let's convert any ``Channel.Alpha`` to ``Color.Blue``.
 
 After providing a default, Chimney can prove the transformation
 is total and use provided function, when it's needed.
+
+Transformations between flat sealed trait hierarchies and deep trait
+hierarchies containing nested sealed traits is also available.
```
Comment (krzemin) [inline docs/source/transformers/customizing-transformers.rst:178]: is -> are

File: chimney/src/test/scala/io/scalaland/chimney/IssuesSpec.scala [hunk 2]
```diff
@@ -491,3 +495,19 @@ object tag {
     def apply[T](t: T): T @@ U = t.asInstanceOf[T @@ U]
   }
 }
+
+object foo {
```
Comment (krzemin) [inline chimney/src/test/scala/io/scalaland/chimney/IssuesSpec.scala:499]: maybe we can have it in `examples` package?

**Comment by krzemin** [comment]
> Maybe allow only transformations, when list of destination parents is prefix of list of source parents and order of parents is linear?

Hmm, I'm not sure if your example causes any issues on deriving a transformer. Well, we can have different inheritance paths, but as long as we match by name and able to construct transformer, it should be fine?

---

I'd be worried about slightly different case:

```scala
object From {
  sealed trait Base
  case object Leaf extends Base
}

object To {
  sealed trait Base
  case object Leaf extends Base
  object Inner {
    case object Leaf extends Base
  }
}
```
Which target leaf would be returned as a result? `To.Leaf` or `To.Inner.Leaf`? I think that currently we don't resolve such conflicts. But that's a separate issue, not really related to this PR. Suggest creating another issue+PR and failing early with derivation error in case we find conflicting names in the target sealed hierarchy.

**Comment by ulanzetz** [comment]
>To.Leaf or To.Inner.Leaf
Nondeterministic. Maybe fail, if more than one subclasses with same name?

**Comment by krzemin** [comment]
> Nondeterministic. Maybe fail, if more than one subclasses with same name?

If they are defined with the same name in the same scope, the compiler just fails with an error. Otherwise (as in the example), it depends on in which order `knownDirectSubclasses` call returns them. Other than that, we are currently deterministic I think. But we can improve on that.

**Review by krzemin — APPROVED** [review]
LGTM 👍

**Comment by ulanzetz** [comment]
@krzemin knownDirectSubclasses returns Set of Symbol, if there are less then 5 different subclasses, list of toInstances will bу ordered and choice will be deterministic. But other it will be different for each compilation
```
object Test {
  object From {
    sealed trait Base

    case object Leaf extends Base
  }

  object To {
    sealed trait Base

    case object Leaf extends Base

    object a {
      case object Leaf extends Base
    }

    object b {
      case object Leaf extends Base
    }

    object c {
      case object Leaf extends Base
    }

    object d {
      case object Leaf extends Base
    }

    object e {
      case object Leaf extends Base
    }
  }

  import io.scalaland.chimney.dsl._

  (From.Leaf: From.Base).transformInto[To.Base]
}
```
I have compiled two times, first time `To.c.Leaf.type` was chosen, second time - `To.b.Leaf.type`

**Comment by krzemin** [comment]
Created issue for that specific improvement: https://github.com/scalalandio/chimney/issues/188

**Comment by kubukoz** [comment]
@krzemin are you planning a release soon? There's a bunch of stuff pending ;)

**Comment by krzemin** [comment]
Hoping for early/mid this week, together with #190 when it's ready.

**Comment by kubukoz** [comment]
Cool, cool, thanks!

**Pull request merged**